File: content-solutions/implementation/sim-snake.js
function solution(n, apples, times, dirs) {
  const dr = [0, 1, 0, -1];
  const dc = [1, 0, -1, 0];
  const key = (r, c) => r * n + c;
  const apple = new Set(apples.map(([r, c]) => key(r, c)));
  const occupied = new Set([key(0, 0)]);
  let d = 0;
  let t = 0;
  let k = 0;
  const cells = [[0, 0]];
  let tail = 0;
  for (;;) {
    t++;
    const [hr, hc] = cells[cells.length - 1];
    const nr = hr + dr[d];
    const nc = hc + dc[d];
    if (nr < 0 || nr >= n || nc < 0 || nc >= n || occupied.has(key(nr, nc))) return t;
    cells.push([nr, nc]);
    occupied.add(key(nr, nc));
    if (apple.has(key(nr, nc))) apple.delete(key(nr, nc));
    else {
      const [tr, tc] = cells[tail++];
      occupied.delete(key(tr, tc));
    }
    if (k < times.length && times[k] === t) {
      d = dirs[k] === "D" ? (d + 1) % 4 : (d + 3) % 4;
      k++;
    }
  }
}

File: content-solutions/implementation/sim-snake.java
import java.util.*;

class Solution {
    static final int[] DR = {0, 1, 0, -1};
    static final int[] DC = {1, 0, -1, 0};

    public int solution(int n, int[][] apples, int[] times, String dirs) {
        Set<Integer> apple = new HashSet<>();
        for (int[] a : apples) apple.add(a[0] * n + a[1]);
        ArrayDeque<int[]> body = new ArrayDeque<>();
        body.addFirst(new int[] {0, 0});
        Set<Integer> occupied = new HashSet<>();
        occupied.add(0);
        int d = 0, t = 0, k = 0;
        while (true) {
            t++;
            int[] head = body.peekFirst();
            int nr = head[0] + DR[d], nc = head[1] + DC[d];
            if (nr < 0 || nr >= n || nc < 0 || nc >= n || occupied.contains(nr * n + nc)) return t;
            body.addFirst(new int[] {nr, nc});
            occupied.add(nr * n + nc);
            if (!apple.remove(nr * n + nc)) {
                int[] tail = body.pollLast();
                occupied.remove(tail[0] * n + tail[1]);
            }
            if (k < times.length && times[k] == t) {
                d = dirs.charAt(k) == 'D' ? (d + 1) % 4 : (d + 3) % 4;
                k++;
            }
        }
    }
}
